src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,216 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as bodyParser from 'body-parser';
+import * as crypto from 'crypto';
+import * as electron from 'electron';
+import * as express from 'express';
+import * as http from 'http';
+import * as path from 'path';
+import * as request from 'request';
+
+const REGISTERED_REDIRECTS = [
+  {clientId: '7f84935771d49c2331e1cfb60c7827e20eaf128103435d82ad20b3c53253b721', port: 55189},
+  {clientId: '4af51205e8d0d8f4a5b84a6b5ca9ea7124f914a5621b6a731ce433c2c7db533b', port: 60434},
+  {clientId: '706928a1c91cbd646c4e0d744c8cbdfbf555a944b821ac7812a7314a4649683a', port: 61437}
+];
+
+function randomValueHex(len: number): string {
+  return crypto.randomBytes(Math.ceil(len / 2))
+      .toString('hex')  // convert to hexadecimal format
+      .slice(0, len);   // return required number of characters
+}
+
+interface ServerError extends Error {
+  code: string;
+}
+
+// Makes server listen on each of the listed ports until there's one open.
+// Returns the index of the port used.
+function listenOnFirstPort(server: http.Server, portList: number[]): Promise<number> {
+  let portIdx = 0;
+  return new Promise((resolve, reject) => {
+    server.once('listening', () => {
+      console.log(`Listening on port ${portList[portIdx]}`);
+      resolve(portIdx);
+    });
+    server.on('error', (error: ServerError) => {
+      if (error.code === 'EADDRINUSE') {
+        console.log(`Port ${portList[portIdx]} already in use`);
+        portIdx += 1;
+        if (portIdx < portList.length) {
+          const port = portList[portIdx];
+          console.log(`Trying port ${port}`);
+          server.listen({host: 'localhost', port, exclusive: true});
+          return;
+        }
+      }
+      server.close();
```

Comment:
Maybe log here.

Reply:
We already do in the handler of the error and on server 'close' event